In [73]:
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt 
import seaborn as sns 

In [74]:
from sklearn.model_selection import train_test_split 
from sklearn.preprocessing import StandardScaler 
from sklearn import datasets
from sklearn.ensemble import RandomForestRegressor

In [75]:
boston = datasets.load_boston() 

In [76]:
boston.feature_names

array(['CRIM', 'ZN', 'INDUS', 'CHAS', 'NOX', 'RM', 'AGE', 'DIS', 'RAD',
       'TAX', 'PTRATIO', 'B', 'LSTAT'], dtype='<U7')

In [77]:
#boston

In [78]:
# X = pd.DataFrame(boston.data, columns=boston.feature_names) 
# y = boston.target 
# y

In [79]:
#X

In [46]:
#boston.DESCR

In [47]:
#boston.data

array([[6.3200e-03, 1.8000e+01, 2.3100e+00, ..., 1.5300e+01, 3.9690e+02,
        4.9800e+00],
       [2.7310e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9690e+02,
        9.1400e+00],
       [2.7290e-02, 0.0000e+00, 7.0700e+00, ..., 1.7800e+01, 3.9283e+02,
        4.0300e+00],
       ...,
       [6.0760e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        5.6400e+00],
       [1.0959e-01, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9345e+02,
        6.4800e+00],
       [4.7410e-02, 0.0000e+00, 1.1930e+01, ..., 2.1000e+01, 3.9690e+02,
        7.8800e+00]])

In [48]:
#features

In [49]:
#targets

In [80]:
bos= pd.DataFrame(boston.data,columns=boston.feature_names)
bos['price']=boston.target

In [81]:
bos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 506 entries, 0 to 505
Data columns (total 14 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   CRIM     506 non-null    float64
 1   ZN       506 non-null    float64
 2   INDUS    506 non-null    float64
 3   CHAS     506 non-null    float64
 4   NOX      506 non-null    float64
 5   RM       506 non-null    float64
 6   AGE      506 non-null    float64
 7   DIS      506 non-null    float64
 8   RAD      506 non-null    float64
 9   TAX      506 non-null    float64
 10  PTRATIO  506 non-null    float64
 11  B        506 non-null    float64
 12  LSTAT    506 non-null    float64
 13  price    506 non-null    float64
dtypes: float64(14)
memory usage: 55.5 KB


In [82]:
bos.head()

CRIM    ZN  INDUS  CHAS    NOX     RM   AGE     DIS  RAD    TAX  \
0  0.00632  18.0   2.31   0.0  0.538  6.575  65.2  4.0900  1.0  296.0   
1  0.02731   0.0   7.07   0.0  0.469  6.421  78.9  4.9671  2.0  242.0   
2  0.02729   0.0   7.07   0.0  0.469  7.185  61.1  4.9671  2.0  242.0   
3  0.03237   0.0   2.18   0.0  0.458  6.998  45.8  6.0622  3.0  222.0   
4  0.06905   0.0   2.18   0.0  0.458  7.147  54.2  6.0622  3.0  222.0   

   PTRATIO       B  LSTAT  price  
0     15.3  396.90   4.98   24.0  
1     17.8  396.90   9.14   21.6  
2     17.8  392.83   4.03   34.7  
3     18.7  394.63   2.94   33.4  
4     18.7  396.90   5.33   36.2

In [83]:
X = bos.iloc[:,:-1]
y = bos.iloc[:,-1]

In [84]:
from sklearn.model_selection import KFold

In [85]:
cv=KFold(n_splits=10,shuffle=True,random_state=2103)

In [86]:
cv

KFold(n_splits=10, random_state=2103, shuffle=True)

In [94]:
list(cv.split(X))

[(array([  0,   1,   2,   3,   4,   5,   6,   7,   9,  10,  11,  12,  13,
          14,  15,  16,  17,  18,  19,  20,  21,  22,  23,  26,  27,  28,
          29,  31,  32,  33,  34,  35,  36,  37,  38,  40,  41,  42,  43,
          44,  45,  46,  47,  48,  49,  50,  51,  52,  53,  54,  56,  57,
          58,  60,  61,  62,  63,  64,  66,  68,  69,  70,  71,  72,  73,
          74,  75,  76,  77,  78,  79,  81,  84,  85,  86,  87,  88,  89,
          90,  91,  92,  93,  94,  95,  96,  97,  98,  99, 100, 101, 103,
         104, 105, 106, 107, 108, 109, 110, 111, 112, 113, 114, 115, 116,
         117, 119, 120, 121, 122, 123, 124, 125, 126, 127, 128, 129, 131,
         132, 133, 134, 135, 136, 137, 138, 140, 141, 142, 143, 144, 145,
         146, 147, 148, 149, 150, 151, 152, 153, 154, 155, 156, 157, 158,
         159, 160, 162, 163, 165, 167, 168, 169, 170, 171, 172, 173, 174,
         176, 177, 178, 179, 180, 181, 183, 184, 185, 187, 188, 189, 190,
         192, 193, 194, 195, 196, 197,

In [95]:
len(list(cv.split(X)))

10

In [96]:
rand_reg = RandomForestRegressor(n_estimators = 10, random_state = 42)
for train_fold, valid_fold in cv.split(X):
    X_train = X.loc[train_fold] # Extract train data with cv indices
    X_valid = X.loc[valid_fold] # Extract valid data with cv indices
    
    y_train = y.loc[train_fold]
    y_valid = y.loc[valid_fold]
    
    model = rand_reg.fit(X = X_train, y = y_train)
    model_pred = model.predict(X = X_valid)
    errors = abs(model_pred - y_valid)
# Print out the mean absolute error
print('Mean Absolute Error:', round(np.mean(errors), 2))

Mean Absolute Error: 2.42


In [97]:
# Print the first 10 true and predicted responses
model_pred[0:10]

array([32.13, 21.68, 25.78, 19.64, 19.62, 19.42, 20.48, 45.65, 21.42,
       23.72])

In [98]:
y_valid[0:10].values

array([26.6, 22.6, 27.5, 19.4, 21.7, 23. , 19.6, 50. , 23.8, 24.6])

In [99]:
# Splitting the dataset into Training set and Test set
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state = 42)

In [100]:
rand_reg = RandomForestRegressor(n_estimators = 10, random_state = 42)
rand_reg.fit(X_train,y_train)

RandomForestRegressor(n_estimators=10, random_state=42)

In [101]:
y_pred = rand_reg.predict(X_test)

In [102]:
y_pred

array([23.49, 30.14, 17.86, 23.32, 16.29, 21.8 , 20.85, 16.54, 20.75,
       19.96, 19.04, 19.59,  8.38, 21.42, 18.51, 22.08, 18.66,  8.82,
       47.08, 15.84, 24.17, 25.98, 14.42, 23.7 , 13.54, 14.17, 21.48,
       14.32, 18.82, 20.99, 19.56, 23.62, 32.77, 20.03, 14.33, 14.87,
       32.92, 18.78, 21.6 , 24.04, 19.77, 29.34, 44.96, 19.69, 23.56,
       13.62, 15.  , 24.67, 18.75, 28.13, 19.45, 33.68, 16.16, 25.18,
       43.58, 22.38, 14.66, 29.71, 22.42, 20.67, 25.93, 33.9 , 26.  ,
       18.12, 25.38, 15.59, 12.92, 23.7 , 28.22, 14.97, 21.12, 27.03,
       10.94, 19.94, 22.38,  7.65, 19.94, 45.05, 11.23, 13.45, 21.1 ,
       14.51, 17.54,  9.97, 20.02, 26.99, 14.81, 23.58, 23.29, 17.49,
       21.03,  7.32, 19.12, 18.97, 25.73, 19.87, 44.39, 14.5 , 12.57,
       14.11, 20.25, 23.85])

In [103]:
# Calculate the absolute errors
errors = abs(y_pred - y_test)
# Print out the mean absolute error
print('Mean Absolute Error:', round(np.mean(errors), 2))

Mean Absolute Error: 2.08


#### Observation : Applying 10 fold cross validation on datapoints and fitting those in random forest regressor, Mean absolute error value is 2.42 as opposed to Mean absolute error value of 2.08 while applying train test split in random forest regressor.

In [50]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test =train_test_split(X,y,random_state=2103)

In [87]:
#from scipy.special import comb

In [52]:
import pandas as pd
#from sklearn.tree import DecisionTreeClassifier
from sklearn.tree import DecisionTreeRegressor
#from sklearn.ensemble import RandomForestClassifier
from sklearn.ensemble import RandomForestRegressor
from sklearn import tree 
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import StandardScaler
#from sklearn.metrics import accuracy_score,confusion_matrix,roc_curve,roc_auc_score
#from sklearn.externals.six  
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error
from sklearn.metrics import r2_score
from IPython.display import Image  
from sklearn.tree import export_graphviz
#import pydotplus

In [53]:
#let's first visualize the tree on the data without doing any pre processing
clf = DecisionTreeRegressor()
clf.fit(X_train,y_train)

DecisionTreeRegressor()

In [54]:
#### We have not done scaling bcoz its a Tree based model

In [55]:
clf.score(X_test,y_test)

0.8136476443940035

In [56]:
rand_reg= RandomForestRegressor(random_state=6)

In [57]:
rand_reg.fit(X_train,y_train)

RandomForestRegressor(random_state=6)

In [58]:
rand_reg.score(X_test,y_test)##returns R Squared

0.8900647368569015

In [59]:
#### We see that the accuracy has hugely improved using Random Forest inplace of Dtree

In [60]:
#### Random forest hyperparameters are a combination of best hyperparameters of both decision tree and Bagging classifier

In [61]:
#### GridSearchCV is a method used to tune our hyperparameters. We can pass different values of hyperparameters as parameters
#### for grid search. It does a exhaustive generation of combination of different parameters passed.
#Using cross validation score, Grid Search returns the combination of hyperparameters for which the model is performing the best.

In [88]:
# # we are tuning three hyperparameters right now, we are passing the different values for both parameters
# grid_param = {
#     "n_estimators" : [90,100,115,130],
#     'criterion': ['mse', 'entropy'],
#     'max_depth' : range(2,20,1),
#     'min_samples_leaf' : range(1,10,1),
#     'min_samples_split': range(2,10,1),
#     'max_features' : ['auto','log2']
# }

In [63]:
#grid_search = GridSearchCV(estimator=rand_reg,param_grid=grid_param,cv=5,n_jobs =-1,verbose = 3)

In [89]:
#(X_train,y_train)

In [90]:
#let's see the best parameters as per our grid search
#grid_search.best_params_

In [ ]:
# # let's save the model

# import pickle

# with open('D:\ineuron_materials_ipynb\iNeuron\EnsembleLearning_And_RandomForest'+ '/modelForPrediction.sav', 'wb') as f:
#     pickle.dump(rand_reg,f)


In [67]:
# rand_reg = RandomForestRegressor(criterion= 'gini',
#  max_features = 'auto',
#  min_samples_leaf = 1,
#  min_samples_split= 2,
#  n_estimators = 90,random_state=6)

In [91]:
#rand_reg(X_train,y_train)

In [69]:
####This error occurs when we miss giving .fit

In [92]:
#rand_reg.fit(X_train,y_train)